### PR TITLE
Acam/place perp order v2

### PR DESCRIPTION
**Pull request by acamill**
Not sure about the destination.

No test, just a first draft @dafyddd let me know if anything sketchy

### CODE REVIEW

File: program/src/matching.rs [hunk 1]
```diff
@@ -271,6 +271,16 @@ pub struct BookSideIter<'a> {
     right: usize,
 }
 
+pub struct Order {
+    // The quantity, in base_lot
+    pub quantity: i64,
+    // Marginal Price, the price to place the order at, in quote (per base_lot)
+    pub price: i64,
+    // The resulting total amount that will be spent, in quote_lot (without fees)
+    pub size: i64,
```
Comment (ckamm) [inline program/src/matching.rs:280]: size currently unused?

File: program/src/processor.rs [hunk 2]
```diff
@@ -2479,6 +2479,167 @@ impl Processor {
...
+                }))
+            }
+            OrderType::ImmediateOrCancel | OrderType::Market => {
+                // Goes through the order book and find the price of quantity for the order
+                let depth_limit = 10;
+                if max_base_quantity > 0 {
+                    book.get_best_order_for_base_quantity(side, max_base_quantity, depth_limit);
+                }
+                if max_quote_quantity > 0 {
+                    book.get_best_order_for_quote_amount(side, max_quote_quantity, depth_limit);
```
Comment (ckamm) [inline program/src/processor.rs:2586]: The return values here and above are unused?

What's the difference between `get_best_order_for_base_quantity()` and using an extreme `limit_price` with the previous order matching?

Is the idea that one can combine `max_base_quantity` and `max_quote_quantity` constraints?

File: program/src/instruction.rs [hunk 3]
```diff
@@ -237,6 +237,31 @@ pub enum MangoInstruction {
...
+    /// `max_base_quantity` will simply be `quantity`, and the
+    /// `limit_price` be `price`.
+    ///
+    /// The max parameters are useful when using instantaneous
+    /// OrderTypes (IoC or Market).
+    /// Goal is to allow orders with thresholds, without knowing
+    /// the quantity nor price in advance.
+    /// It will check wether it fits within the provided `limit_price`,
+    /// It will optionally checks wether it's fully filled for IoC.
+    PlacePerpOrderV2 {
```
Comment (ckamm) [inline program/src/instruction.rs:253]: note: I'll also make a PlacePerpOrderV2 for the time-in-force feature. Let's get both new features into that new instruction.

File: program/src/matching.rs (hunk 1 above)
Comment (acamill) [reply program/src/matching.rs:280]: Yes size is not used anymore, can be removed

File: program/src/processor.rs [hunk 4]
```diff
@@ -2479,6 +2479,167 @@ impl Processor {
...
+                    price: limit_price,
+                    size: 0,
+                    side,
+                }))
+            }
+            OrderType::ImmediateOrCancel | OrderType::Market => {
+                // Goes through the order book and find the price of quantity for the order
+                let depth_limit = 10;
+                if max_base_quantity > 0 {
+                    book.get_best_order_for_base_quantity(side, max_base_quantity, depth_limit);
```
[the remark this answers is not in the document]
Comment (acamill) [reply program/src/processor.rs:2583]: I was trying to do a first port of existing logic and then refine along the way. It's still very rough, but yes can probably fit there better.

File: program/src/processor.rs (hunk 2 above)
Comment (acamill) [reply program/src/processor.rs:2586]: No these constraints are independants. 

On my side I was using these two function to calculate an order before it happen to see how much collateral need to be deposited on mango and estimate other things.

I then saw the interface from 01 add these parameters in they equivalent of "place_perp_order" so I gave it a go, but it's probably gibberish as it is.

**Comment by acamill** [comment]
Code deprecated from our side, and @ckamm working on implementing the max_quote_lot version after discussions today

**Pull request closed without merging**